cli/command/stack/kubernetes/list.go

```diff
@@ -44,6 +54,55 @@ func getStacks(kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error)
 	return formattedStacks, nil
 }
 
+func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error) {
+	stacks, err := getStacks(kubeCli, opts)
+	if err == nil || !apierrs.IsForbidden(err) {
+		return stacks, err
+	}
+	nms, err2 := getUserVisibleNamespaces(dockerCli)
+	if err2 != nil {
+		logrus.Debugf("Failed to query user visible namespaces: %s", err2)
+		return nil, err
+	}
+	opts.AllNamespaces = false
+	for _, namespace := range nms.Items {
+		kubeCli.kubeNamespace = namespace.Name
+		ss, err := getStacks(kubeCli, opts)
+		if err != nil {
+			return nil, err
+		}
+		stacks = append(stacks, ss...)
+	}
+	return stacks, nil
+}
+
+func getUserVisibleNamespaces(dockerCli command.Cli) (*core_v1.NamespaceList, error) {
+	host := dockerCli.Client().DaemonHost()
+	endpoint, err := url.Parse(host)
+	if err != nil {
+		return nil, err
+	}
+	endpoint.Scheme = "https"
+	endpoint.Path = "/kubernetesNamespaces"
+	resp, err := dockerCli.Client().HTTPClient().Get(endpoint.String())
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "received %d status and unable to read response", resp.StatusCode)
+	}
+	if resp.StatusCode != http.StatusOK {
```

Comment:
Yes, but you just log a debug message there and return `err`.

Reply:
It logs `err2`, the error returned by `getUserVisibleNamespaces`, and then returns `err` from the previous call, which is different and in practice due to the `apierrs.IsForbidden(err)` test right before is effectively as if `getUserVisibleNamespaces` had not been called at all:
```
$ docker --orchestrator=all stack ls
stacks.compose.docker.com is forbidden: User "user" cannot list stacks.compose.docker.com at the cluster scope: access denied
```

Reply:
I see. I think a comment here would be helpful.

Reply:
OK, I added
```
		// ignore this last error and return the previous one instead
```

Reply:
So, thinking:

- A 404 error is not an error, just means: we don't have UCP
- We're now hiding potential other errors (500 - UCP crashed, 403 - UCP is up, but credentials are incorrect (?), ??? - Other error)

Instead of returning expected errors, and ignoring them in `getStacksWithAllNamespaces()`, we should leave the logic/decision if it's an actual error or not to `getUserVisibleNamespaces()`. When running into a 404, just return an empty list.

```go
func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts options.List) ([]*formatter.Stack, error) {
	stacks, err := getStacks(kubeCli, opts)
	if err == nil || !apierrs.IsForbidden(err) {
		return stacks, err
	}
	nms, err2 := getUserVisibleNamespaces(dockerCli)
	if err2 != nil {
		return nil, errors.Wrap(err2, "Failed to query user visible namespaces")
	}
	if err != nil {
		return nil, errors.Wrap(err, "Failed to retrieve stacks")
	}
	opts.AllNamespaces = false
	for _, namespace := range nms.Items {
		kubeCli.kubeNamespace = namespace.Name
		ss, err := getStacks(kubeCli, opts)
		if err != nil {
			return nil, err
		}
		stacks = append(stacks, ss...)
	}
	return stacks, nil
}

func getUserVisibleNamespaces(dockerCli command.Cli) (*core_v1.NamespaceList, error) {
	host := dockerCli.Client().DaemonHost()
	endpoint, err := url.Parse(host)
	if err != nil {
		return nil, err
	}
	endpoint.Scheme = "https"
	endpoint.Path = "/kubernetesNamespaces"
	resp, err := dockerCli.Client().HTTPClient().Get(endpoint.String())
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		return nil, errors.Wrapf(err, "received %d status and unable to read response", resp.StatusCode)
	}
	namespaces := &core_v1.NamespaceList{}
	switch resp.StatusCode {
	case http.StatusOK:
		if err := json.Unmarshal(body, namespaces); err != nil {
			return nil, errors.Wrapf(err, "unmarshal failed: %s", string(body))
		}
	case http.StatusNotFound:
		// UCP API not present
	default:
		return nil, fmt.Errorf("received %d status while retrieving namespaces: %s", resp.StatusCode, string(body))
	}
	return namespaces, nil
}
```


```patch
diff --git a/cli/command/stack/kubernetes/list.go b/cli/command/stack/kubernetes/list.go
index 8a7b1406..cb467f9b 100644
--- a/cli/command/stack/kubernetes/list.go
+++ b/cli/command/stack/kubernetes/list.go
@@ -11,7 +11,6 @@ import (
        "github.com/docker/cli/cli/command/formatter"
        "github.com/docker/cli/cli/command/stack/options"
        "github.com/pkg/errors"
-       "github.com/sirupsen/logrus"
        core_v1 "k8s.io/api/core/v1"
        apierrs "k8s.io/apimachinery/pkg/api/errors"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -61,9 +60,10 @@ func getStacksWithAllNamespaces(dockerCli command.Cli, kubeCli *KubeCli, opts op
        }
        nms, err2 := getUserVisibleNamespaces(dockerCli)
        if err2 != nil {
-               logrus.Debugf("Failed to query user visible namespaces: %s", err2)
-               // ignore this last error and return the previous one instead
-               return nil, err
+               return nil, errors.Wrap(err2, "Failed to query user visible namespaces")
+       }
+       if err != nil {
+               return nil, errors.Wrap(err, "Failed to retrieve stacks")
        }
        opts.AllNamespaces = false
        for _, namespace := range nms.Items {
@@ -94,12 +94,16 @@ func getUserVisibleNamespaces(dockerCli command.Cli) (*core_v1.NamespaceList, er
        if err != nil {
                return nil, errors.Wrapf(err, "received %d status and unable to read response", resp.StatusCode)
        }
-       if resp.StatusCode != http.StatusOK {
-               return nil, fmt.Errorf(string(body))
-       }
        namespaces := &core_v1.NamespaceList{}
-       if err := json.Unmarshal(body, namespaces); err != nil {
-               return nil, errors.Wrapf(err, "unmarshal failed: %s", string(body))
+       switch resp.StatusCode {
+       case http.StatusOK:
+               if err := json.Unmarshal(body, namespaces); err != nil {
+                       return nil, errors.Wrapf(err, "unmarshal failed: %s", string(body))
+               }
+       case http.StatusNotFound:
+               // UCP API not present
+       default:
+               return nil, fmt.Errorf("received %d status while retrieving namespaces: %s", resp.StatusCode, string(body))
        }
        return namespaces, nil
 }
```